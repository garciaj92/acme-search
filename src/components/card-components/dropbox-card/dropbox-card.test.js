import React from "react";
import DropboxCard from "./dropbox-card";

import { mount, shallow } from "enzyme";
import toJson from 'enzyme-to-json';

// Mocks
import { consoleErrorFn } from "../../../__mocks__/console.mock";

const props = {
    title: "Dropbox Test",
    path: "/test/path/dropbox.json",
    sharedWith: ["me@jorgegarcia.co"],
    created: "01/01/2020 - 4:00:00 PM"
};

describe("DropboxCard: Test visual components", () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<DropboxCard {...props} />);
    });

    it("renders correctly", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        // console.log(wrapper.debug());
    });

    it("displays the props correctly", () => {
        expect(wrapper.find("[data-testid='card-heading']").first().text()).toEqual(props.title);
        expect(wrapper.find("[data-testid='path-info'] h5").first().text()).toEqual("Path");
        expect(wrapper.find("[data-testid='path-info'] p").first().text()).toEqual(props.path);
        expect(wrapper.find("[data-testid='shared-with-info'] h5").first().text()).toEqual("Shared with");
        expect(wrapper.find("[data-testid='shared-with-info'] a").first().text()).toEqual(props.sharedWith[0]);
        expect(wrapper.find("[data-testid='card-footer'] h5").first().text()).toEqual("Created");
        expect(wrapper.find("[data-testid='card-footer'] p").first().text()).toEqual(props.created);
    });
});


describe("DropboxCard: Test prop type validation", () => {
    let consoleErrorSpy;
    beforeEach(() => {
        console.error = consoleErrorFn
        consoleErrorSpy = jest.spyOn(console, "error");
    });

    afterEach(() => {
        consoleErrorSpy.mockClear();
    });

    it("does not throw an error when props are provided", () => {
        shallow(<DropboxCard {...props} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    it("does not throw an error when sharedWith prop is not provided", () => {
        const { title, path, created } = props;
        shallow(<DropboxCard title={title} path={path} created={created} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(0);
    });

    // What is stated below applies to the remaining test cases in this describe block:
    //      console.error is called twice for each prop validator that fails because
    //      there are two components that use each failing prop. This means that if
    //      two props fail we have four errors.

    it("throws an error when title prop is not of type `string`", () => {
        const { path, created } = props;
        shallow(<DropboxCard title={1} path={path} created={created} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when path prop is not of type `string`", () => {
        const { title, created } = props;
        shallow(<DropboxCard title={title} path={1} created={created} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when created prop is not of type `string`", () => {
        const { title, path } = props;
        shallow(<DropboxCard title={title} path={path} created={1} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });

    it("throws an error when sharedWith prop is not an arrays of type `string`", () => {
        const { title, path, created } = props;
        shallow(<DropboxCard title={title} path={path} created={created} sharedWith={[0,1]} />);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(2);
    });
});