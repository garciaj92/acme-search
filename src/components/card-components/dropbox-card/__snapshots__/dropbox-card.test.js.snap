// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropboxCard: Test visual components renders correctly 1`] = `
<DropboxCard
  created="01/01/2020 - 4:00:00 PM"
  path="/test/path/dropbox.json"
  sharedWith={
    Array [
      "me@jorgegarcia.co",
    ]
  }
  title="Dropbox Test"
>
  <Card
    bg="white"
    body={false}
    className="shadow calendar-card"
  >
    <div
      className="shadow calendar-card card bg-white"
    >
      <CardBody>
        <div
          className="card-body"
        >
          <CardHeading
            data-testid="card-heading"
            heading="Dropbox Test"
          >
            <h4
              className="font-weight-bold"
            >
              Dropbox Test
            </h4>
          </CardHeading>
          <PathInfo
            data-testid="path-info"
            path="/test/path/dropbox.json"
          >
            <CardIconSubHeading
              heading="Path"
              iconSrc="file.svg"
            >
              <div
                className="d-flex justify-content-center align-items-center my-2"
              >
                <Image
                  fluid={false}
                  rounded={false}
                  roundedCircle={false}
                  src="file.svg"
                  thumbnail={false}
                  width={30}
                >
                  <img
                    className=""
                    src="file.svg"
                    width={30}
                  />
                </Image>
                <h5
                  className="mb-0 mx-1"
                >
                  Path
                </h5>
              </div>
            </CardIconSubHeading>
            <CardLink
              content="/test/path/dropbox.json"
            >
              <p
                className="text-muted mb-0 d-block"
              >
                /test/path/dropbox.json
              </p>
            </CardLink>
          </PathInfo>
          <SharedWithInfo
            data-testid="shared-with-info"
            emails={
              Array [
                "me@jorgegarcia.co",
              ]
            }
          >
            <CardIconSubHeading
              heading="Shared with"
              iconSrc="people-outline.svg"
            >
              <div
                className="d-flex justify-content-center align-items-center my-2"
              >
                <Image
                  fluid={false}
                  rounded={false}
                  roundedCircle={false}
                  src="people-outline.svg"
                  thumbnail={false}
                  width={30}
                >
                  <img
                    className=""
                    src="people-outline.svg"
                    width={30}
                  />
                </Image>
                <h5
                  className="mb-0 mx-1"
                >
                  Shared with
                </h5>
              </div>
            </CardIconSubHeading>
            <div
              className="d-flex flex-wrap justify-content-center"
            >
              <CardLink
                content="me@jorgegarcia.co"
                href="mailto:me@jorgegarcia.co"
                key="0"
              >
                <a
                  className="text-muted mb-0 d-block"
                  href="mailto:me@jorgegarcia.co"
                  onClick={[Function]}
                >
                  me@jorgegarcia.co
                </a>
              </CardLink>
            </div>
          </SharedWithInfo>
        </div>
      </CardBody>
      <CardFooter
        className="bg-white"
        data-testid="card-footer"
      >
        <div
          className="bg-white card-footer"
          data-testid="card-footer"
        >
          <DateInfo
            date="01/01/2020 - 4:00:00 PM"
            title="Created"
          >
            <div
              className="text-muted"
            >
              <h5
                className="font-weight-bold mb-0"
              >
                Created
              </h5>
              <p
                className="font-weight-bold mb-0"
              >
                01/01/2020 - 4:00:00 PM
              </p>
            </div>
          </DateInfo>
        </div>
      </CardFooter>
    </div>
  </Card>
</DropboxCard>
`;
